src/Momento.Sdk/Responses/TopicMessage.cs

```diff
@@ -0,0 +1,93 @@
+using Momento.Protos.CacheClient.Pubsub;
+using Momento.Sdk.Exceptions;
+
+namespace Momento.Sdk.Responses;
+
+/// <summary>
+/// Parent type for a topic message. The message is resolved to a type-safe
+/// object of one of the following subtypes:
+/// <list type="bullet">
+/// <item><description>TopicMessage.Item</description></item>
+/// <item><description>TopicMessage.Error</description></item>
+/// </list>
+/// Pattern matching can be used to operate on the appropriate subtype.
+/// For example:
+/// <code>
+/// if (message is TopicMessage.Item item)
+/// {
+///     return item.ValueString();
+/// }
+/// else if (message is TopicMessage.Error error)
+/// {
+///     // handle error as appropriate
+/// }
+/// else
+/// {
+///     // handle unexpected response
+/// }
+/// </code>
+/// </summary>
+public abstract class TopicMessage
+{
+    /// <summary>
+    /// A topic message containing a value. If the value is a string, ValueString
+    /// will return it. If the value is binary, ValueByteArray will return it.
+    /// </summary>
+    public class Item : TopicMessage
+    {
+        private readonly _TopicValue _value;
+
+        /// <summary>
+        /// Constructs an Item from an internal _TopicItem
+        /// </summary>
+        /// <param name="topicItem">Containing the binary or string value.</param>
+        public Item(_TopicItem topicItem)
+        {
+            _value = topicItem.Value;
+            TopicSequenceNumber = topicItem.TopicSequenceNumber;
+        }
```

Comment:
so, oneofs in protocol buffers are a little different for each language, but they always expect some kind of matching before accessing a property. As-is, a user can't tell the difference between a published empty string and a binary message.
We should provide a way for users to be able to tell the difference, possibly similar to the way protocol buffers does it (that's where we got the go matching pattern), between an Error, a Discontinuity, a Text value and a Binary value. Here is the documentation for inspiration: https://learn.microsoft.com/en-us/dotnet/architecture/grpc-for-wcf-developers/protobuf-any-oneof#oneof

```
        public Item(_TopicItem topicItem)
        {
            switch (topicItem.Value.KindCase)
            {
                case _TopicValue.KindOneofCase.Text:
                    {
                        // The only populated field in this message's value is Text.
                        // Let's vend it to the user that way instead of leaving them to guess?
                        break;
                    }
                case _TopicValue.KindOneofCase.Binary:
                    {
                        // The only populated field in this message's value is Binary.
                        break;
                    }   
                case _TopicValue.KindOneofCase.None:
                    {
                        // Skip message? It's not useful, and the server did something bad
                        break;
                    }
                default:
                    {
                        // Skip message? It's not useful, as it holds something undefined
                        break;
                    }
            }
            _value = topicItem.Value;
            TopicSequenceNumber = topicItem.TopicSequenceNumber;
        }
```

Reply:
I agree with that. I was trying to follow the pattern set by the cache methods, where the value is derived from the same ByteString regardless of whether it is text or binary. Since the difference is more explicit here I'm in favor of making them separate types.

Reply:
@kvcache so do you really think that's the right API here? To force users to always have a switch statement in order to read anything at all off of the subscription, even for the simplest use cases?

that feels overbearing and in the other SDKs I think we have been trying to avoid forcing that on them.

Reply:
@kvcache random thought:

In the JS SDK we implement subscriptions by allowing the user to pass callbacks for OnItem and OnError. Presumably in the future there might also be OnDiscontinuity.

We did that in JS SDK because that's how the grpc streaming API was exposed.

In golang we did it iterator-style because that's how the grpc streaming API was exposed.

But would it be crazy, in .NET, to do it the callback way?

We could encapsulate the foreach loop inside the SDK in an async function, and allow the user to pass us callbacks for OnBinary, OnText, OnDiscontinuity.

Then they wouldn't have to have a switch statement.

It seems like it might be a better API IMO, with two questions in my mind:

1. Is it going to be obvious to them that the overloaded signatures of `publish` will map to the specific `OnBinary`/`OnText` callbacks? (This issue is relevant now, even if we stick with the iterator-style API.)
2. If they need to do anything that has to do with ordering between OnText and OnDiscontinuity, having those separated into two user-provided lambdas might make it harder for them to manage/share state across the two callbacks? maybe not though, they could close over it or use a class that had the two callback functions?

Reply:
@cprice404 this is inside the Item class - I do think it is reasonable to expect Momento developers to switch to do the needful with each message on the subscription :-).

Separately, I do not think all users should be forced to use pattern matching if they are of the exceptional persuasion. I tried to suggest that we should have accessors available that throw if the message is not of the required disposition. I think the item type should BOTH [1] have a switch pattern available AND [2] have exceptional typed accessors available.

Reply:
oh dude, you sparked a memory: I think we might want to think about using `delegate`s for this. You hit on something interesting here.

Reply:
```c#
class SubscriptionActions
{
    public SubscriptionActions(TextHandler? onText = null, BinaryHandler? onBinary = null)
    {
        OnTextMessage += Log;
        OnBinaryMessage += Log;

        if (onText != null)
        {
            OnTextMessage += onText;
        }
        if (onBinary != null)
        {
            OnBinaryMessage += onBinary;
        }
    }

    public event TextHandler OnTextMessage;
    public event BinaryHandler OnBinaryMessage;

    private void Log<T>(T message)
    {
    }
}

class Lol
{
    SubscriptionActions MakeSubscriptionActionsThatJustPrintTextMessagesToTheConsole()
    {
        return new SubscriptionActions(onText: Console.WriteLine);
    }
}
```

Reply:
Hmm, as nice as this _could_ be, I think `await foreach` is too ergonomic to ignore.